real_preprocessing/src/opt_visualization.cpp

```diff
@@ -0,0 +1,220 @@
+#include "ros/ros.h"
+#include "ros/package.h"
+#include "fstream"
+#include "string"
+#include "yaml-cpp/yaml.h"
+#include "opencv2/opencv.hpp"
+#include "Eigen/Dense"
+#include "visualization_msgs/Marker.h"
+#include "visualization_msgs/MarkerArray.h"
+#include "geometry_msgs/Pose.h"
+#include "geometry_msgs/PoseArray.h"
+
+//Returns the absolute directory path for package detections folder
+std::string getBasePath()
+{
+	return (ros::package::getPath("real_preprocessing") +"/detections");
+}
+
+//Returns the absolute directory path for detections files
+std::string getDetfilePath(int filenum)
+{
+	return (getBasePath() + "/detections_" + std::to_string(filenum) + ".yaml");
+}
+
+//Returns the absolute directory path for targets yaml file
+std::string getTargetPath()
+{
+	return (getBasePath() + "/targets.yaml");
+}
+
+//converts and returns a quaternion given a rodrigues angle
+Eigen::Quaterniond rodrigToQuat(double rod_vec[3])
+{
+	cv::Mat rodrigues=cv::Mat(1,3,CV_64F,rod_vec);
+	cv::Mat rot_mat;
+	cv::Rodrigues(rodrigues,rot_mat);
+	Eigen::Matrix3d rot;
+	rot << rot_mat.at<double>(0, 0), rot_mat.at<double>(0, 1), rot_mat.at<double>(0, 2),
+		   rot_mat.at<double>(1, 0), rot_mat.at<double>(1, 1), rot_mat.at<double>(1, 2), 
+	       rot_mat.at<double>(2, 0), rot_mat.at<double>(2, 1), rot_mat.at<double>(2, 2);
+	Eigen::Quaterniond quat(rot);
+	return quat;
+}
+
+//generates a pose message given a tag index in the targets.yaml file
+geometry_msgs::Pose loadMarkerPose(int tag_index)
+{
+	geometry_msgs::Pose pose_msg;
+	YAML::Node YAML_handle= YAML::LoadFile(getTargetPath());
+	
+	pose_msg.position.x = YAML_handle["targets"][tag_index]["world_T_target"]["translation"][0].as<double>();
+	pose_msg.position.y = YAML_handle["targets"][tag_index]["world_T_target"]["translation"][1].as<double>();
+	pose_msg.position.z = YAML_handle["targets"][tag_index]["world_T_target"]["translation"][2].as<double>();
+
+	double rod_vec[3] = {YAML_handle["targets"][tag_index]["world_T_target"]["rotation"][0].as<double>(),
+						 YAML_handle["targets"][tag_index]["world_T_target"]["rotation"][1].as<double>(),
+						 YAML_handle["targets"][tag_index]["world_T_target"]["rotation"][2].as<double>()};
+		 
+	Eigen::Quaterniond quat = rodrigToQuat(rod_vec);
+	pose_msg.orientation.x = quat.x();
+	pose_msg.orientation.y = quat.y();
+	pose_msg.orientation.z = quat.z();
+	pose_msg.orientation.w = quat.w();
+
+	return pose_msg;
+}
+
+//populates a tag marker message with a world_T_tag pose of the indexed tag in the targets.yaml file 
```

Comment:
Populate a tag marker and tag pose message, given a tag_index to load the world_T_tag pose within targets.yaml file.